codebasin/config.py

```diff
@@ -212,150 +216,161 @@ class ClangCompiler(Compiler):
-            defines.append("_OPENMP")
+            if pass_ == "default" and omp:
+                defines.append("_OPENMP")
 
-        return defines
+            configuration = PreprocessorConfiguration(
+                defines,
+                include_paths,
+                include_files,
+                pass_name=pass_,
+            )
 
+            configurations.append(configuration)
 
-class GnuCompiler(Compiler):
+        return configurations
+
+
+class GnuArgumentParser(ArgumentParser):
     """
     Represents the behavior of GNU-based compilers.
     """
 
-    def __init__(self, argv: list[str]):
-        super().__init__(argv)
-
+    def parse_args(self, argv: list[str]) -> list[PreprocessorConfiguration]:
+        args = _parse_compiler_args(argv + _importcfg[self.name])
         for arg in argv:
             if arg in ["-fopenmp"]:
-                self.defines.append("_OPENMP")
-                break
-
-
-class HipCompiler(Compiler):
-    """
-    Represents the behavior of the HIP compiler.
-    """
-
-    def __init__(self, argv: list[str]):
-        super().__init__(argv)
-
-
-class IntelCompiler(ClangCompiler):
-    """
-    Represents the behavior of Intel compilers.
-    """
-
-    def __init__(self, argv: list[str]):
-        super().__init__(argv)
+                args.defines.append("_OPENMP")
+        configuration = PreprocessorConfiguration(
+            args.defines,
+            args.include_paths,
+            args.include_files,
+        )
+        return [configuration]
 
 
-class NvccCompiler(Compiler):
+class NvccArgumentParser(ArgumentParser):
     """
     Represents the behavior of the NVCC compiler.
     """
 
-    def __init__(self, argv: list[str]):
-        super().__init__(argv)
-        self.omp = False
+    def parse_args(self, argv: list[str]) -> list[PreprocessorConfiguration]:
+        args = _parse_compiler_args(argv + _importcfg[self.name])
+
+        omp = False
+        passes = {"default"}
 
         for arg in argv:
             archs = re.findall("sm_(\\d+)", arg)
             archs += re.findall("compute_(\\d+)", arg)
-            self.passes |= set(archs)
+            passes |= set(archs)
 
             if arg in ["-fopenmp", "-fiopenmp", "-qopenmp"]:
-                self.omp = True
+                omp = True
                 continue
 
-    def get_defines(self, pass_):
-        defines = super().get_defines(pass_)
+        configurations = []
+        for pass_ in passes:
+            defines = args.defines.copy()
+            include_files = args.include_files.copy()
+            include_paths = args.include_paths.copy()
 
-        defines.append("__NVCC__")
-        defines.append("__CUDACC__")
+            defines.append("__NVCC__")
+            defines.append("__CUDACC__")
 
-        if pass_ != "default":
-            arch = int(pass_) * 10
-            defines.append(f"__CUDA_ARCH__={arch}")
+            if pass_ != "default":
+                arch = int(pass_) * 10
```

Comment:
Leave a comment to explain what the `* 10` is for?

Reply:
Good idea. Done in 37b567530fb3e4a56b69bf33679624d99275b141.